### DISCUSSION TITLE
YOLOV8's TensorRT speeds up reasoning and asks for help

### DISCUSSION THREAD

**Discussion by hua-ever** [ultralytics/ultralytics#8994] [Q&A]
Dear friends,
I have now encountered a problem. I have implemented the deployment of yolov5 model on tensorRT, but with the update of my project, the V5 model is no longer enough to meet the requirements. I am seeking to modify the existing YOLOV5 deployment code so that it can use the YOLOV8 model. I have looked for a lot of materials and tried, but still failed, now I am looking for help, if you can give me some help to modify the code, I would be very grateful.

**Comment by pderrenger** [comment]
Hello there! 😊

We're excited to hear you're considering YOLOv8 for your project! While migrating from YOLOv5 to YOLOv8 for TensorRT deployments involves slight modifications, the core process remains very similar. Here’s a quick guide to get you started:

1. **Export YOLOv8 Model to ONNX**: First, export your YOLOv8 model to ONNX format. You can use the `yolo export` command for this.
    ```bash
    yolo export model=yolov8n.pt format=onnx
    ```
2. **Convert ONNX to TensorRT**: Use the `onnx2trt` tool or TensorRT APIs to convert your ONNX model to a TensorRT engine. Remember to adjust parameters based on your YOLOv8 model specifics.
    ```bash
    onnx2trt yolov8n.onnx -o yolov8n.trt
    ```
3. **Update Inference Code**: You might need to slightly adjust the preprocessing or postprocessing steps in your code due to differences between YOLOv5 and YOLOv8 architectures. Pay close attention to input size and output format.

If specific errors or issues arise during this process, feel free to share them here. The YOLO community and the Ultralytics team are here to help! 🚀

Best of luck with your YOLOv8 deployment, and don't hesitate to reach out for further assistance.

**Reply by hua-ever** [reply]
![20240322-213720](https://github.com/ultralytics/ultralytics/assets/148219004/ee9ccca6-9c51-4d11-8f59-1a71055e8f85)
Thanks for your help before, now I have modified the code and made the V8 model run successfully, but now there is a serious problem, just like the situation in the picture above, the number and position of the boundary box for target recognition in the picture is not ideal, what I recognize is the number plate in the picture, normally, it should only have one box. And it is just the size of the box number, please help me solve this problem

**Reply by pderrenger** [reply]
@hua-ever hey there! 🌟

Great to hear you've got the V8 model up and running! 🚀 Regarding the issue with multiple bounding boxes around the license plate, it seems like an instance of overlapping detections. A quick tweak you could try is adjusting the confidence (`conf`) and Non-Maximum Suppression (`iou`) thresholds to filter out weaker, overlapping detections. Here's how you might adjust those settings:

```python
from ultralytics import YOLO

# Load your YOLOv8 model
model = YOLO('yolov8n.pt')

# Adjust confidence and IoU thresholds
results = model.predict(source='path/to/your/image.jpg', conf=0.5, iou=0.5)
```

Increasing the `conf` value makes the model more conservative, only showing detections it's more certain about. Adjusting the `iou` helps in handling overlapping detections more aggressively.

Give it a try, and let's see how it goes! If you're still facing issues, we're here to dive deeper and assist further. 🛠️

**Reply by hua-ever** [reply]
Thank you very much for your help, but I think I have a new problem.Please allow me to ask you for help again.The following piece of code is the part where I post-process the object recognition output.
But it doesn't help me process the result correctly. It turns out that no matter how I change the identification object, the result he gives is the same and in the same place. I looked for a lot of data and found that maybe there was a problem with the way I got the output information after processing. 
In the post-processing part, I used the method of obtaining x,y,wight,high and confidence of yoloV5, but I could not modify this part correctly, so that it could run correctly. I hope you can give me some suggestions to help me modify this code, which may be the key to my final success

```
void TRTInfer::decode_output(std::vector<DetectionObj> &res, cv::Mat &frame, float *pdata, float &obj_threshold, float &confidence_threshold, float &nms_threshold)
{   

    int n = 0, q = 0, i = 0, j = 0, row_ind = 0, k = 0;
    for (n = 0; n < this->num_stride; ++n)
    {
        const float stride = pow(2, n + 3); // calculate stride
        int num_grid_x = (int)ceil(this->input_dims.d[3] / stride); 
        int num_grid_y = (int)ceil(this->input_dims.d[2] / stride);

        // Analytic model output
        for (q = 0; q < 3; ++q) // ergodic anchor box
        {
            for (i = 0; i < num_grid_y; ++i)
            {
                for (j = 0; j < num_grid_x; ++j)
                {
                    // Resolve bounding box coordinates and confidence
                    float *pitem = pdata + row_ind * this->output_dims.d[1]; // A pointer to the current bounding box
                    float box_score = pitem[4]; // 边界框置信度
                    if (box_score > obj_threshold) 
                    {
                        int max_ind = 0;
                        float max_class_score = 0;
                        for (k = 0; k < this->num_classes; ++k)
                        {
                            float class_score = pitem[k + 5];
                            if (class_score > max_class_score)
                            {
                                max_class_score = class_score;
                                max_ind = k;
                            }
                        }
                        max_class_score *= box_score; 
                        if (max_class_score > confidence_threshold)
                        {   
                            float cx = pitem[0]; //  x 
                            float cy = pitem[1]; //  y 
                           
                            float w = pitem[2];  // wight
                            float h = pitem[3];  // high
                        }
                    }
                    ++row_ind;
                }
            }
        }
    }
```

**Reply by pderrenger** [reply]
Hi there! 😊

It looks like you're on the right track with your code, but transitioning from YOLOv5 to YOLOv8 requires some adjustments, especially in post-processing due to potential differences in output formats.

In YOLOv8, the prediction output generally includes the bounding box coordinates, confidence scores, and class probabilities directly. However, the way these outputs are interpreted and scaled might differ slightly. Make sure your output processing aligns with how YOLOv8 structures its output. If you're using TensorRT, it's crucial to ensure the outputs you're reading match those expected by your post-processing step.

Here's a simplified pattern to follow when processing your outputs. This doesn't directly fit into your code but provides a conceptual guide:

```cpp
float conf = pitem[4]; // Confidence
if (conf > obj_threshold) {
    int class_id = -1;
    float max_class_score = -1;

    // Find class with highest score
    for (int class_index = 0; class_index < num_classes; ++class_index) {
        float class_score = pitem[5 + class_index];
        if (class_score > max_class_score) {
            max_class_score = class_score;
            class_id = class_index;
        }
    }

    float final_score = conf * max_class_score;
    if (final_score > confidence_threshold) {
        // Bounding Box info
        float bx = pitem[0];
        float by = pitem[1];
        float bw = pitem[2];
        float bh = pitem[3];

        // Convert center coordinates to top left format if necessary
        // float x = bx - bw / 2;
        // float y = by - bh / 2;
    }
}
```

Remember, this is a generic guide. Adapt it based on how YOLOv8's output tensor is structured in your specific model version and deployment framework.

Checking your decoding loop index calculation (especially with strides and grid sizes), and ensuring you're correctly iterating through the output tensor will be crucial. Without knowing the exact format of YOLOv8’s TensorRT output tensor, I’d recommend double-checking the model’s documentation or reaching out directly on GitHub discussions for specifics tailored to YOLOv8. 

Hope this points you in the right direction! 🚀
